### PR TITLE
Add support to retrieve kernel cache (retrieve notify)

### PR DESCRIPTION
**Pull request by navytux**
This patch continues bdca0e6a (Add support for store notify) and adds
corresponding support for counterpart to cache-store - to retrieve an inode
data from kernel cache.

As it was already noted in bdca0e6a, FUSE protocol provides primitives for
pagecache control: to invalidate a data region for inode (notify_inval_inode),
to store data into inode kernel's cache (notify_store), and to retrieve data
from inode kernel's cache (notify_retrieve). For the latter 2 FUSE protocol
messages and brief documentation about semantic can be seen here:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/fuse.h?id=v4.19-rc6-177-gcec4de302c5f#n68
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/fuse.h?id=v4.19-rc6-177-gcec4de302c5f#n756

https://git.kernel.org/linus/a1d75f2582
https://git.kernel.org/linus/2d45ba381a

In short, to retrieve data from kernel cache, filesystem server sends

	S > C	NOTIFY_RETRIEVE_CACHE{notifyUnique, inode, offset, size}

and if that message was sent correctly, the kernel sends back another
write-style message with unique=notifyUnique

	S < C	NOTIFY_REPLY{inode, offset, size, data}

Since so far there were no cases when a server was querying the kernel,
and the reply comes as separate kernel "request", we have to add
infrastructure for tracking such in-flight queries. This is done by
adding Server.retrieveTab and friends.

Otherwise the implementation is straightforward.

A particular note is that from a user-level API point of view we are not
following e.g. libfuse to register a callback to be invoked upon reply, but
instead provide {Inode,File}RetrieveCache that synchronously send notify query
and wait for kernel's reply.

This fits more naturally to Go and is easier to use.

### CODE REVIEW

**Comment by navytux** [comment]
@hanwen, would you please have a look?

Thanks beforehand,
Kirill

**Comment by hanwen** [comment]
I am taking holidays. Can it wait for another week?
On Mon, Dec 3, 2018 at 9:37 AM Kirill Smelkov <notifications@github.com> wrote:
>
> @hanwen, would you please have a look?
>
> Thanks beforehand,
> Kirill
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.



-- 
Han-Wen Nienhuys - hanwenn@gmail.com - http://www.xs4all.nl/~hanwen

**Comment by navytux** [comment]
Yes, sure. Thanks for feedback. I did not knew you were away and it was just a gentle ping. It of course can wait until you are back.

File: fuse/opcode.go [hunk 1]
```diff
@@ -177,6 +178,45 @@ func doWrite(server *Server, req *request) {
...
+		badf('E', "offset mismatch: expected @%d, got @%d", reading.offset, reply.Offset)
+		return
+	}
+
+	if len(reading.dest) < len(req.arg) {
+		badf('W', "too much data: requested [%d], got [%d]", len(reading.dest), len(req.arg))
+	}
+
+	// XXX copy -> TODO splice
+	reading.n  = copy(reading.dest, req.arg)
```
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/opcode.go]: Yes, that's why I put the TODO - for when we'll have the support to rework this place.

File: fuse/test/cachecontrol_test.go [hunk 2]
```diff
@@ -95,6 +95,39 @@ func TestCacheControl(t *testing.T) {
...
+			n, st := fsconn.FileRetrieveCache(file.Inode(), offset, buf)
+			if st != fuse.OK {
+				t.Fatalf("%s: retrieve cache @%d [%d]: %s", subj, offset, size, st)
+			}
+			if got := buf[:n]; string(got) != dataOK {
+				t.Fatalf("%s: retrieve cache @%d [%d]: have %q; want %q", subj, offset, size, got, dataOK)
+			}
+		}
+
+		// retrieve [1:-1]   (also verifying that offset/size are handled correctly)
```
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/test/cachecontrol_test.go]: It's python notation. There `A[-X]` is shorthand for `A[len(A)-X]` and similarly for slices. In other words `-X` means Xth element from the end (starting counting from the offset just past the array).

The code here does actually check for about

```go
l := len(dataOK)
dataOK[1:l-1]
```

but I find it more clear to describe in overview with `[1:-1]` in the for-human comment.

Probably it is my Python background that shows, though I hope it is ok.

File: fuse/nodefs/fsconnector.go [hunk 3]
```diff
@@ -407,6 +407,26 @@ func (c *FileSystemConnector) FileNotifyStoreCache(node *Inode, off int64, data
 	return c.server.InodeNotifyStoreCache(nID, off, data)
 }
 
+// FileRetrieveCache retrieves data from kernel's inode cache.
+//
+// This call retrieves data from kernel's inode cache @ offset and up to
+// len(dest) bytes. If kernel cache has fewer consequtive data starting at
+// offset, that fewer amount is returned. In particular if inode data at offset
+// is not cached (0, OK) is returned.
```
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/nodefs/fsconnector.go]: Hm, we have this pattern in many places in nodefs: when a function accepts an Inode and tries to convert the Inode into integer node ID, and if it does not find the handle for the Inode it returns EINVAL. For example here:

https://github.com/hanwen/go-fuse/blob/7954c216/fuse/nodefs/fsconnector.go#L422
https://github.com/hanwen/go-fuse/blob/7954c216/fuse/nodefs/fsconnector.go#L440
https://github.com/hanwen/go-fuse/blob/7954c216/fuse/nodefs/fsconnector.go#L404

and none of this functions mention EINVAL in their docstring.

Actually isn't it a _bug_, if `Inode` is used but is not in the handle map?
Shouldn't in that case it thus should be not EINVAL, but panic instead?

I agree that similar functions at raw fuse level might return EINVAL abour wrong node ID, but that EINVAL will come from kernel in that case. However at nodefs level there should not be case when a valid Inode is used and it is not in `inodeMap`. Or am I missing something?

Given the above what we should do for my `nodefs.FileRetrieveCache(Inode, ...)`? Document that if invalid Inode is passed in EINVAL will be returned? or we should just conver all the above cases to panics?

P.S. 

https://github.com/hanwen/go-fuse/blob/7954c216/fuse/nodefs/fsconnector.go#L381 returns even OK if the Inode was not found in inodeMap. The rationale here is, probably, that since there is nothing in inode map, the kernel does not know about this inode, and then there is no need to notify it about that the inode content has been changed. But still, isn't it a bug?

File: fuse/test/cachecontrol_test.go [hunk 4]
```diff
@@ -95,6 +95,39 @@ func TestCacheControl(t *testing.T) {
...
+		if l >= 2 {
+			assertCacheReadAt(1, l-2, dataOK[1:l-1])
+		}
+
+		// retrieve [:∞]
+		assertCacheReadAt(0, l+10000, dataOK)
+	}
+
+	// lookup on the file - forces to assign inode ID to it
+	// XXX better do it without kernel - how?
```
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/test/cachecontrol_test.go]: We could register node ID to inode when it is created in `nodefs.newInode`. This is completely inside filesystem server and well defined. Now the client - the kernel - won't know that node ID before it actually tries to access corresponding file. And all cache operations on that inode should be likely noop: on retrieve it is ok to say the client has nothing, and on store most of the time it is probably ok to say ok, since the kernel is free to drop cached data. The only case is when user asked the kernel to mlock corresponding pages, but then the node ID should be already know to the kernel and thus to the handle map.

Should I modify `nID == 0` handling in store/retrieve cache according to the above logic?
Should we modify nodefs logic to automatically assign node ID right immediately when an inode is created?

If so, I think this stat can be dropped from the test.

File: fuse/opcode.go [hunk 5]
```diff
@@ -177,6 +178,45 @@ func doWrite(server *Server, req *request) {
...
 }
 
+func doNotifyReply(server *Server, req *request) {
+	reply := (*NotifyRetrieveIn)(req.inData)
+	server.retrieveMu.Lock()
+	reading := server.retrieveTab[reply.Unique]
+	delete(server.retrieveTab, reply.Unique)
+	server.retrieveMu.Unlock()
+
+	badf := func(level byte, format string, argv ...interface{}) {
```
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/opcode.go]: Could you please clarify? The whole `doNotifyReply` is newly introduced in this patch. When we have a notify reply message from the kernel there are various cases that could be wrong, e.g. if it was some mismatch between what kernel sent and what we are expecting to receive. If that is the case we should somehow report the problem - log it. When logging from under notify reply handler there are several log messages possible, but they all have the same prefix. I just factored such logging with the same prefix into internal to doNotifyReply function. The level is also there because one badness condition is not too bad - e.g. if the kernel sent us more data than we requested, we can ignore the tail but still use the requested part. Still it was some error logic, probably on the kernel side, to send more data, and it is better for it to not to go unnoticed. That's why badf also accepts the prefix.

In other words here it is mostly syntatic sugar not to duplicate the error prefix.

Do you want me to remove badf from this patch and just use log.Printf directly?

**Comment by navytux** [comment]
@hanwen, thanks for feedback. I replied to your comments and there is one question about inode ID and how/when they are creaed. Let's please try to sort them out.

About logging - I have large changes to overhaul logging, so that it is more signal/noise in fuse message, comparred to current situation, and though I do not insist on that change being accepted, I will try to post it here just in case it could be maybe useful.

**Comment by navytux** [comment]
Log SNR improvements: https://github.com/hanwen/go-fuse/pull/249.

**Comment by navytux** [comment]
rebased on top of master to solve minor conflict + gofmt.

---- 8< ---- (interdiff)
```diff
--- a/fuse/opcode.go
+++ b/fuse/opcode.go
@@ -186,7 +186,7 @@ func doNotifyReply(server *Server, req *request) {
        server.retrieveMu.Unlock()
 
        badf := func(level byte, format string, argv ...interface{}) {
-               log.Printf(string(level) + ": notify reply: " + format, argv...)
+               log.Printf(string(level)+": notify reply: "+format, argv...)
        }
 
        if reading == nil {
@@ -194,7 +194,7 @@ func doNotifyReply(server *Server, req *request) {
                return
        }
 
-       reading.n  = 0
+       reading.n = 0
        reading.st = EIO
        defer close(reading.ready)
 
@@ -213,7 +213,7 @@ func doNotifyReply(server *Server, req *request) {
        }
 
        // XXX copy -> TODO splice
-       reading.n  = copy(reading.dest, req.arg)
+       reading.n = copy(reading.dest, req.arg)
        reading.st = OK
 }
 
diff --git a/fuse/server.go b/fuse/server.go
index f07f727..d526d5d 100644
--- a/fuse/server.go
+++ b/fuse/server.go
@@ -347,7 +347,7 @@ func (ms *Server) Serve() {
        ms.retrieveTab = make(map[uint64]*retrieveCacheRequest)
        ms.retrieveMu.Unlock()
        for _, reading := range rtab {
-               reading.n  = 0
+               reading.n = 0
                reading.st = ENODEV
                close(reading.ready)
        }
```

**Comment by navytux** [comment]
( CI is flaking in unionfs on go1.9 )

File: fuse/nodefs/fsconnector.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hanwen) [reply fuse/nodefs/fsconnector.go]: if there is no inode number, we can't talk to the kernel, so I think there is nothing to be done. EINVAL isn't documented in other places, but maybe those other places should be fixed?

The other Notify functions can return OK, because if the kernel doesn't know about the inode, then its cache can certainly not be out of date.

File: fuse/opcode.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hanwen) [reply fuse/opcode.go]: I'm talking about just the logging. I introduces log levels (errors, warning) with abbreviations (E, W), which aren't used in other places. I think it's good to introduce some more structure into the log messages, but would rather see a change that addresses all log statements throughout the code.

File: fuse/opcode.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hanwen) [reply fuse/opcode.go]: I have no plans to add splicing for that direction.

File: fuse/test/cachecontrol_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hanwen) [reply fuse/test/cachecontrol_test.go]: I know python notation, but Go is not python, so better to just phrase it as [1:len - 1] instead.

File: fuse/nodefs/fsconnector.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/nodefs/fsconnector.go]: Ok, thanks for clarifying. I also thought a bit and looked at fuse inside kernel a bit and now it all makes sense: if the kernel did not lookup on the inode, or it did the lookup but later forgot, it does not have an entry for the inode in its dentry cache and so directly talking to kernel with any notify operation about the inode will return `ENODEV`.

Now for operations that have semantic to just invalidate the cache, we can pretend that they succeed. For `FileRetrieveCache` we can also pretend that the cache is empty and reads as just (0, OK) instead of EINVAL.

For `FileNotifyStoreCache` I indeed have to fix the docstring and the implementation to return ENOENT for the case when the kernel does not have a handle for the inode. And it was the only case where we were returning EINVAL.

I will amend the whole change with the following:

```diff
diff --git a/fuse/nodefs/fsconnector.go b/fuse/nodefs/fsconnector.go
index 5010dd8..2a94319 100644
--- a/fuse/nodefs/fsconnector.go
+++ b/fuse/nodefs/fsconnector.go
@@ -393,6 +393,9 @@ func (c *FileSystemConnector) FileNotify(node *Inode, off int64, length int64) f
 // and will use data from that cache for non direct-IO reads from the inode
 // in corresponding data region. After kernel's cache data is evicted, the kernel
 // will have to issue new Read calls on user request to get data content.
+//
+// ENOENT is returned if the kernel does not currently have entry for this
+// inode in its dentry cache.
 func (c *FileSystemConnector) FileNotifyStoreCache(node *Inode, off int64, data []byte) fuse.Status {
        var nID uint64
        if node == c.rootNode {
@@ -402,7 +405,8 @@ func (c *FileSystemConnector) FileNotifyStoreCache(node *Inode, off int64, data
        }
 
        if nID == 0 {
-               return fuse.EINVAL
+               // the kernel does not currently know about this inode.
+               return fuse.ENOENT
        }
        return c.server.InodeNotifyStoreCache(nID, off, data)
 }
@@ -422,7 +426,9 @@ func (c *FileSystemConnector) FileRetrieveCache(node *Inode, off int64, dest []b
        }
 
        if nID == 0 {
-               return 0, fuse.EINVAL
+               // the kernel does not currently know about this inode.
+               // -> we can pretend that its cache for the inode is empty.
+               return 0, fuse.OK
        }
        return c.server.InodeRetrieveCache(nID, off, dest)
 }
diff --git a/fuse/test/cachecontrol_test.go b/fuse/test/cachecontrol_test.go
index 78cba41..5c1bc84 100644
--- a/fuse/test/cachecontrol_test.go
+++ b/fuse/test/cachecontrol_test.go
@@ -121,8 +121,15 @@ func TestCacheControl(t *testing.T) {
                assertCacheReadAt(0, l+10000, dataOK)
        }
 
+       // before the kernel has entry for file in its dentry cache, the cache
+       // should read as empty and store should fail with ENOENT.
+       assertCacheRead("before lookup", "")
+       st := fsconn.FileNotifyStoreCache(file.Inode(), 0, []byte("abc"))
+       if st != fuse.ENOENT {
+               t.Fatalf("%s: store cache -> %v; want %v", "before lookup", st, fuse.ENOENT)
+       }
+
        // lookup on the file - forces to assign inode ID to it
-       // XXX better do it without kernel - how?
        os.Stat(dir + "/hello.txt")
 
        // cache should be initially empty
@@ -168,7 +175,7 @@ func TestCacheControl(t *testing.T) {
        assertCacheRead("original", data0)
 
        // store changed data into OS cache
-       st := fsconn.FileNotifyStoreCache(file.Inode(), 7, []byte("123"))
+       st = fsconn.FileNotifyStoreCache(file.Inode(), 7, []byte("123"))
        if st != fuse.OK {
                t.Fatalf("store cache: %s", st)
        }
```

File: fuse/test/cachecontrol_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/test/cachecontrol_test.go]: Please see https://github.com/hanwen/go-fuse/pull/246#discussion_r242149177. I'm dropping this XXX.

File: fuse/opcode.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/opcode.go]: Ok, I'm dropping log levels here.

File: fuse/opcode.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/opcode.go]: Ok, let's drop this XXX then.

File: fuse/test/cachecontrol_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (navytux) [reply fuse/test/cachecontrol_test.go]: Ok, changed to `[1:len-1]`.

**Comment by navytux** [comment]
Thanks for feedback. I have amended the change to address all review comments. Now there are 2 patches here - one for cache retrieve, and another one that fixes cache store to return `ENOENT` when inode is not known to kernel.

I hope it is ok.

Cumulative interdiff is as follows:

```diff
diff --git a/fuse/nodefs/fsconnector.go b/fuse/nodefs/fsconnector.go
index 5010dd8..7717a1d 100644
--- a/fuse/nodefs/fsconnector.go
+++ b/fuse/nodefs/fsconnector.go
@@ -393,6 +393,9 @@ func (c *FileSystemConnector) FileNotify(node *Inode, off int64, length int64) f
 // and will use data from that cache for non direct-IO reads from the inode
 // in corresponding data region. After kernel's cache data is evicted, the kernel
 // will have to issue new Read calls on user request to get data content.
+//
+// ENOENT is returned if the kernel does not currently have entry for this
+// inode in its dentry cache.
 func (c *FileSystemConnector) FileNotifyStoreCache(node *Inode, off int64, data []byte) fuse.Status {
 	var nID uint64
 	if node == c.rootNode {
@@ -402,7 +405,8 @@ func (c *FileSystemConnector) FileNotifyStoreCache(node *Inode, off int64, data
 	}
 
 	if nID == 0 {
-		return fuse.EINVAL
+		// the kernel does not currently know about this inode.
+		return fuse.ENOENT
 	}
 	return c.server.InodeNotifyStoreCache(nID, off, data)
 }
@@ -410,9 +414,13 @@ func (c *FileSystemConnector) FileNotifyStoreCache(node *Inode, off int64, data
 // FileRetrieveCache retrieves data from kernel's inode cache.
 //
 // This call retrieves data from kernel's inode cache @ offset and up to
-// len(dest) bytes. If kernel cache has fewer consequtive data starting at
+// len(dest) bytes. If kernel cache has fewer consecutive data starting at
 // offset, that fewer amount is returned. In particular if inode data at offset
 // is not cached (0, OK) is returned.
+//
+// If the kernel does not currently have entry for this inode in its dentry
+// cache (0, OK) is still returned, pretending that the inode could be known to
+// the kernel, but kernel's inode cache is empty.
 func (c *FileSystemConnector) FileRetrieveCache(node *Inode, off int64, dest []byte) (n int, st fuse.Status) {
 	var nID uint64
 	if node == c.rootNode {
@@ -422,7 +430,9 @@ func (c *FileSystemConnector) FileRetrieveCache(node *Inode, off int64, dest []b
 	}
 
 	if nID == 0 {
-		return 0, fuse.EINVAL
+		// the kernel does not currently know about this inode.
+		// -> we can pretend that its cache for the inode is empty.
+		return 0, fuse.OK
 	}
 	return c.server.InodeRetrieveCache(nID, off, dest)
 }
diff --git a/fuse/opcode.go b/fuse/opcode.go
index 483c11a..7121fdd 100644
--- a/fuse/opcode.go
+++ b/fuse/opcode.go
@@ -185,12 +185,12 @@ func doNotifyReply(server *Server, req *request) {
 	delete(server.retrieveTab, reply.Unique)
 	server.retrieveMu.Unlock()
 
-	badf := func(level byte, format string, argv ...interface{}) {
-		log.Printf(string(level)+": notify reply: "+format, argv...)
+	badf := func(format string, argv ...interface{}) {
+		log.Printf("notify reply: "+format, argv...)
 	}
 
 	if reading == nil {
-		badf('E', "unexpected unique - ignoring")
+		badf("unexpected unique - ignoring")
 		return
 	}
 
@@ -199,20 +199,19 @@ func doNotifyReply(server *Server, req *request) {
 	defer close(reading.ready)
 
 	if reading.nodeid != reply.NodeId {
-		badf('E', "inode mismatch: expected %s, got %s", reading.nodeid, reply.NodeId)
+		badf("inode mismatch: expected %s, got %s", reading.nodeid, reply.NodeId)
 		return
 	}
 
 	if reading.offset != reply.Offset {
-		badf('E', "offset mismatch: expected @%d, got @%d", reading.offset, reply.Offset)
+		badf("offset mismatch: expected @%d, got @%d", reading.offset, reply.Offset)
 		return
 	}
 
 	if len(reading.dest) < len(req.arg) {
-		badf('W', "too much data: requested [%d], got [%d]", len(reading.dest), len(req.arg))
+		badf("too much data: requested %db, got %db (will use only %db)", len(reading.dest), len(req.arg), len(reading.dest))
 	}
 
-	// XXX copy -> TODO splice
 	reading.n = copy(reading.dest, req.arg)
 	reading.st = OK
 }
diff --git a/fuse/server.go b/fuse/server.go
index d526d5d..1ca2854 100644
--- a/fuse/server.go
+++ b/fuse/server.go
@@ -565,6 +565,9 @@ func (ms *Server) inodeNotifyStoreCache32(node uint64, offset int64, data []byte
 // [offset:offset+len(dest)) and waits for corresponding reply. If kernel cache
 // has fewer consecutive data starting at offset, that fewer amount is returned.
 // In particular if inode data at offset is not cached (0, OK) is returned.
+//
+// The kernel returns ENOENT if it does not currently have entry for this inode
+// in its dentry cache.
 func (ms *Server) InodeRetrieveCache(node uint64, offset int64, dest []byte) (n int, st Status) {
 	if !ms.kernelSettings.SupportsNotify(NOTIFY_RETRIEVE_CACHE) {
 		return 0, ENOSYS
@@ -614,16 +617,16 @@ func (ms *Server) InodeRetrieveCache(node uint64, offset int64, dest []byte) (n
 	}
 	if result != OK {
 		ms.retrieveMu.Lock()
-		switch ms.retrieveTab[q.NotifyUnique] {
-		case reading:
+		r := ms.retrieveTab[q.NotifyUnique]
+		if r == reading {
 			delete(ms.retrieveTab, q.NotifyUnique)
-		case nil:
+		} else if r == nil {
 			// ok - might be dequeued by umount
-		default:
-			// although very unlikely, it is possible that kernel sents
+		} else {
+			// although very unlikely, it is possible that kernel sends
 			// unexpected NotifyReply with our notifyUnique, then
 			// retrieveNext wraps, makes full cycle, and another
-			// retrive request is made with the same notifyUnique.
+			// retrieve request is made with the same notifyUnique.
 			log.Printf("W: INODE_RETRIEVE_CACHE: request with notifyUnique=%d mutated", q.NotifyUnique)
 		}
 		ms.retrieveMu.Unlock()
diff --git a/fuse/test/cachecontrol_test.go b/fuse/test/cachecontrol_test.go
index 78cba41..ca73b84 100644
--- a/fuse/test/cachecontrol_test.go
+++ b/fuse/test/cachecontrol_test.go
@@ -111,7 +111,7 @@ func TestCacheControl(t *testing.T) {
 			}
 		}
 
-		// retrieve [1:-1]   (also verifying that offset/size are handled correctly)
+		// retrieve [1:len - 1]   (also verifying that offset/size are handled correctly)
 		l := len(dataOK)
 		if l >= 2 {
 			assertCacheReadAt(1, l-2, dataOK[1:l-1])
@@ -121,8 +121,15 @@ func TestCacheControl(t *testing.T) {
 		assertCacheReadAt(0, l+10000, dataOK)
 	}
 
+	// before the kernel has entry for file in its dentry cache, the cache
+	// should read as empty and cache store should fail with ENOENT.
+	assertCacheRead("before lookup", "")
+	st := fsconn.FileNotifyStoreCache(file.Inode(), 0, []byte("abc"))
+	if st != fuse.ENOENT {
+		t.Fatalf("%s: store cache -> %v; want %v", "before lookup", st, fuse.ENOENT)
+	}
+
 	// lookup on the file - forces to assign inode ID to it
-	// XXX better do it without kernel - how?
 	os.Stat(dir + "/hello.txt")
 
 	// cache should be initially empty
@@ -168,7 +175,7 @@ func TestCacheControl(t *testing.T) {
 	assertCacheRead("original", data0)
 
 	// store changed data into OS cache
-	st := fsconn.FileNotifyStoreCache(file.Inode(), 7, []byte("123"))
+	st = fsconn.FileNotifyStoreCache(file.Inode(), 7, []byte("123"))
 	if st != fuse.OK {
 		t.Fatalf("store cache: %s", st)
 	}
```

**Comment by navytux** [comment]
( CI was falky in unionfs tests on go1.10 )

**Comment by navytux** [comment]
Thanks for merging.
